processor/src/main/java/org/mapstruct/ap/internal/processor/AnnotationBasedComponentModelProcessor.java

```diff
@@ -88,34 +117,119 @@ protected void adjustDecorator(Mapper mapper) {
+        }
+
+        List<Annotation> mapperReferenceAnnotations = getMapperReferenceAnnotations();
+
+        removeReferenceAnnotations( mapperReferencesForConstructor, mapperReferenceAnnotations );
+
+        return new AnnotatedConstructor(
+            mapper.getName(),
+            mapperReferencesForConstructor,
+            mapperReferenceAnnotations,
+            additionalPublicEmptyConstructor() );
+    }
+
+    private AnnotatedConstructor buildAnnotatedConstructorForDecorator(Decorator decorator) {
+        List<AnnotationMapperReference> mapperReferencesForConstructor = new ArrayList<AnnotationMapperReference>();
+
+        for ( Field field : decorator.getFields() ) {
+            if ( field instanceof AnnotationMapperReference ) {
+                mapperReferencesForConstructor.add( (AnnotationMapperReference) field );
+            }
+        }
+
+        List<Annotation> mapperReferenceAnnotations = getMapperReferenceAnnotations();
+
+        removeReferenceAnnotations( mapperReferencesForConstructor, mapperReferenceAnnotations );
+
+        return new AnnotatedConstructor(
+            decorator.getName(),
+            mapperReferencesForConstructor,
+            mapperReferenceAnnotations,
+            additionalPublicEmptyConstructor() );
+    }
+
+    /**
+     * Removes constructor parameter annotations that are already on the constructor. For example, for CDI, the
+     * javax.inject.Inject annotation is on the constructor and must not be on the constructor parameters.
+     *
+     * @param annotationMapperReferences list of annotations that are be on the constructor parameter
+     * @param mapperReferenceAnnotations annotations that are be on the constructor
+     */
+    private void removeReferenceAnnotations(List<AnnotationMapperReference> annotationMapperReferences,
+                                            List<Annotation> mapperReferenceAnnotations) {
+        ListIterator<AnnotationMapperReference> mapperReferenceIterator = annotationMapperReferences.listIterator();
+
+        Set<Type> mapperReferenceAnnotationsTypes = new HashSet<Type>();
+        for ( Annotation annotation : mapperReferenceAnnotations ) {
+            mapperReferenceAnnotationsTypes.add( annotation.getType() );
+        }
+
+        while ( mapperReferenceIterator.hasNext() ) {
+            AnnotationMapperReference annotationMapperReference = mapperReferenceIterator.next();
+            mapperReferenceIterator.remove();
+
+            List<Annotation> qualifiers = new ArrayList<Annotation>();
+            for ( Annotation annotation : annotationMapperReference.getAnnotations() ) {
+                if ( !mapperReferenceAnnotationsTypes.contains( annotation.getType() ) ) {
+                    qualifiers.add( annotation );
+                }
+            }
+
+            mapperReferenceIterator.add(
+                new AnnotationMapperReference(
```

Comment:
I would extract a new method on the `AnnotationMapperReference` where you can pass the new annotations and you get a new instance (in order to reduce this class a bit).

It can be something like:

```java
public AnnotationMapperReference withNewAnnotations(List<Annotation> annotations) {
        return new AnnotationMapperReference(
            getType(),
            getVariableName(),
            annotations,
            isUsed(),
            fieldFinal,
            includeAnnotationsOnField
        );
}
```